ipaserver/plugins/idp.py

```diff
@@ -0,0 +1,304 @@
+            ),
+        Password('ipaidpclientsecret?',
+                 cli_name='secret',
+                 label=_('Secret'),
+                 doc=_('The client secret'),
+                 confirm=True,
+                 flags={'no_display'},
+                 ),
+        Str('ipaidpscope?',
+            cli_name='scope',
+            label=_('Scope'),
+            doc=_('Scope of the access request'),
+            ),
+    )
+
+    permission_filter_objectclasses = ['ipaidp']
+    managed_permissions = {
+        'System: Add External IdP server': {
+            'ipapermright': {'add'},
+            'ipapermlocation': DN(container_dn, api.env.basedn),
+            'ipapermtargetfilter': {
+                '(objectclass=ipaidp)'},
+            'default_privileges': {'External Idp server Administrators'}
+        },
+        'System: Read External IdP server': {
+            'ipapermright': {'read', 'search', 'compare'},
+            'ipapermdefaultattr': {
+                'cn', 'objectclass', 'ipaidpauthendpoint',
+                'ipaidptokenendpoint', 'ipaidpclientid',
+                'ipaidpscope'
+            },
+            'ipapermlocation': DN(container_dn, api.env.basedn),
+            'ipapermtargetfilter': {
+                '(objectclass=ipaidp)'},
+            'default_privileges': {'External Idp server Administrators'}
+        },
+        'System: Modify External IdP server': {
+            'ipapermright': {'write'},
+            'ipapermlocation': DN(container_dn, api.env.basedn),
+            'ipapermdefaultattr': {
+                'cn', 'objectclass', 'ipaidpauthendpoint',
+                'ipaidptokenendpoint', 'ipaidpclientid',
+                'ipaidpscope', 'ipaidpclientsecret',
+            },
+            'default_privileges': {'External Idp server Administrators'}
+        },
+        'System: Delete External IdP server': {
+            'ipapermright': {'delete'},
+            'ipapermlocation': DN(container_dn, api.env.basedn),
+            'ipapermtargetfilter': {
+                '(objectclass=ipaidp)'},
+            'default_privileges': {'External Idp server Administrators'}
+        },
+        'System: Read External IdP server client secret': {
+            'ipapermright': {'read', 'search', 'compare'},
+            'ipapermlocation': DN(container_dn, api.env.basedn),
+            'ipapermdefaultattr': {
+                'cn', 'objectclass', 'ipaidpauthendpoint',
+                'ipaidptokenendpoint', 'ipaidpclientid',
+                'ipaidpscope', 'ipaidpclientsecret'
+            },
+            'ipapermtargetfilter': {
+                '(objectclass=ipaidp)'},
+        }
+    }
+
+
+@register()
+class idp_add(LDAPCreate):
+    __doc__ = _('Add a new Identity Provider server.')
+    msg_summary = _('Added Identity Provider server "%(value)s"')
+
+    # List of pre-populated idp endpoints
+    # key = provider, value = tuple of (auth-uri, token-uri)
+    idp_providers = {
+        u'google': (
```

Comment:
you don't need to use `u'...'` with Python 3 anymore, just use normal strings here and in other places, please.